### PR TITLE
Fix bug in cmp with equality when operand is (p-1)/2 + add condtional_enforce_cmp

### PR DESCRIPTION
**Pull request by nicholas-mainardi**
The comparison functions for `FpGadget` only accepts operands in the range [0, (p-1)/2] for a prime field Fp.
In case equality between the input operands x,y must be also checked in the comparison function (e.g, x <= y), the current implementation perform the comparison by comparing x and y+1 (i.e., to check if x <= y, it is checked if x < y+1). However, if y==(p-1)/2, then y+1 > (p-1)/2, which is not properly handled by the comparison function, hence triggering a constraint violation (although y=(p-1)/2 is a legitimate value according to specs of the comparison functions).

This PR fixes this bug by dealing with the equality case without performing such troublesome addition, exploiting these 2 identities:
- x <= y iff !(x > y)
- x >= y iff !(x < y)

In addition, unit tests are enriched to take into account such corner cases.

Furthermore, this PR adds the comparison functions `conditional_enforce_cmp` and `conditional_enforce_cmp_unchecked`, which enforce an order relationship between the operands if and only if a boolean condition is verified

### CODE REVIEW

File: r1cs/gadgets/std/src/fields/cmp.rs [hunk 1]
```diff
@@ -189,93 +184,188 @@ mod test {
...
+                    _ => {}
                 }
-                Ordering::Greater => {
-                    a_var.enforce_cmp(cs.ns(|| "enforce greater"), &b_var, Ordering::Greater, false).unwrap();
-                    a_var.enforce_cmp(cs.ns(|| "enforce greater equal"), &b_var, Ordering::Greater, true).unwrap();
+
+                if i == 0 {
+                    println!("number of constraints: {}", cs.num_constraints());
+                }
+                if !cs.is_satisfied(){
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/cmp.rs]: Can we remove it from the test ?

File: r1cs/gadgets/std/src/fields/cmp.rs [hunk 2]
```diff
@@ -189,93 +184,188 @@ mod test {
...
-        for _i in 0..10 {
+            // test corner case when operands are extreme values of range [0, (p-1)/2] of
+            // admissible values
             let mut cs = TestConstraintSystem::<Fr>::new();
-            let a = rand_in_range(&mut rng);
-            let a_var = FpGadget::<Fr>::alloc(&mut cs.ns(|| "generate_a"), || Ok(a)).unwrap();
-            a_var.enforce_cmp(cs.ns(|| "enforce less"),&a_var, Ordering::Less, false).unwrap();
+            let max_val: Fr = Fr::modulus_minus_one_div_two().into();
+            let max_var = FpGadget::<Fr>::alloc(&mut cs.ns(|| "generate_max"), || Ok(max_val)).unwrap();
+            let zero_var = FpGadget::<Fr>::from_value(cs.ns(|| "alloc zero"), &Fr::zero());
```
Comment (95DDB) [inline r1cs/gadgets/std/src/fields/cmp.rs]: There is also the zero() function that does the same

**Review by 95DDB — APPROVED** [review]

**Review by 95DDB — APPROVED** [review]
